### PR TITLE
[WIP] New resource aws_db_proxy_target

### PR DESCRIPTION
**Pull request by gazoakley**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates #12690

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
**New Resource:** `aws_db_proxy_target` (#12784)
```

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccAWSDBProxyTarget_'

...
```

### CODE REVIEW

**Comment by teamterraform** [comment]
## Notification of Recent and Upcoming Changes to Contributions

Thank you for this contribution! There have been a few recent development changes that affect this pull request. We apologize for the inconvenience, especially if there have been long review delays up until now. Please note that this is automated message from an unmonitored account. See the [FAQ](https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/FAQ.md) for additional information on the maintainer team and review prioritization.

If you are unable to complete these updates, please leave a comment for the community and maintainers so someone can potentially continue the work. The maintainers will encourage other contributors to use the existing contribution as the base for additional changes as appropriate. Otherwise, contributions that do not receive updated code or comments from the original contributor may be closed in the future so the maintainers can focus on active items.

For the most up to date information about Terraform AWS Provider development, see the [Contributing Guide](https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/CONTRIBUTING.md). Additional technical debt changes can be tracked with the [`technical-debt` label on issues](https://github.com/terraform-providers/terraform-provider-aws/issues?q=is%3Aopen+is%3Aissue+label%3Atechnical-debt).

As part of updating a pull request with these changes, the most current unit testing and linting will run. These may report issues that were not previously reported.

### Terraform 0.12 Syntax

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/8950
Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14417

Version 3 and later of the Terraform AWS Provider, which all existing contributions would potentially be added, only supports Terraform 0.12 and later. Certain syntax elements of Terraform 0.11 and earlier show deprecation warnings during runs with Terraform 0.12. Documentation and test configurations, such as those including deprecated string interpolations (`some_attribute = "${aws_service_thing.example.id}"`) should be updated to the newer syntax (`some_attribute = aws_service_thing.example.id`). Contribution testing will automatically fail on older syntax in the near future. Please see the referenced issues for additional information.

### Action Required: Terraform Plugin SDK Version 2

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14551

The Terraform AWS Provider has been upgraded to the latest version of the Terraform Plugin SDK. Generally, most changes to contributions should only involve updating Go import paths in source code files. Please see the referenced issue for additional information.

### Action Required: Removal of website/aws.erb File

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14712

Any changes to the `website/aws.erb` file are no longer necessary and should be removed from this contribution to prevent merge issues in the near future when the file is removed from the repository. Please see the referenced issue for additional information.

### Upcoming Change of Git Branch Naming

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14292

Development environments will need their upstream Git branch updated from `master` to `main` in the near future. Please see the referenced issue for additional information and scheduling.

### Upcoming Change of GitHub Organization

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14715

This repository will be migrating from https://github.com/terraform-providers/terraform-provider-aws to https://github.com/hashicorp/terraform-provider-aws. No practitioner or developer action is anticipated and most GitHub functionality will automatically redirect to the new location. Go import paths including `terraform-providers` can remain for now. Please see the referenced issue for additional information and scheduling.

**Comment by bryantbiggs** [comment]
@bflad any chance we can get this in 3.6.0 - the aws_rds_proxy resource is not usuable without the target group unfortuantely

**Comment by bflad** [comment]
@gazoakley since the `aws_db_proxy_default_target_group` resource is not required to implement this resource (e.g. you can hardcode `target_group_name = "default"` in the testing, it'll still get the right dependency ordering from `db_proxy_name = aws_db_proxy.test.name`), we can separately test and work through this review if you want by rebasing those changes out of this. 👍 

@bryantbiggs it should be usable to manage any RDS DB Proxy itself 😉 , but yes it is understood that end-to-end management is not yet possible (that would be the closing of the tracking issue #12690). Since this is a community driven effort we cannot provide a guarantee on timeframes unless the contributor because unresponsive and the maintainers take over. The note above would speed things up here though. Larger functionality in this project is always broken up into more iterative pieces so we can work through reviews and release functionality faster than releasing it all at once.

File: website/docs/r/db_proxy_target.html.markdown [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+  require_tls            = true
+  role_arn               = "arn:aws:iam:us-east-1:123456789012:role/example"
+  vpc_security_group_ids = ["sg-12345678901234567"]
+  vpc_subnet_ids         = ["subnet-12345678901234567"]
+
+  auth {
+    auth_scheme = "SECRETS"
+    description = "example"
+    iam_auth    = "DISABLED"
+    secret_arn  = "arn:aws:secretsmanager:us-east-1:123456789012:secret:example"
```
Comment (bflad) [inline website/docs/r/db_proxy_target.html.markdown:30]: Example attribute references:

```suggestion
  role_arn               = aws_iam_role.example.arn
  vpc_security_group_ids = [aws_security_group.example.id]
  vpc_subnet_ids         = [aws_subnet.example.id]

  auth {
    auth_scheme = "SECRETS"
    description = "example"
    iam_auth    = "DISABLED"
    secret_arn  = aws_secretsmanager_secret.example.arn
```

**Review by bflad — CHANGES_REQUESTED** [review]
Just some relatively minor items then should be good to go, thank you @gazoakley 😄

**Comment by swagsb** [comment]
Hi,

When will this be released ?

**Comment by bflad** [comment]
Hi @swagsb 👋  There is no definitive release date yet. We will update the pull request milestone if/when there is a specific release. Please do note though that this pull request is covered by the [August - October 2020 Roadmap](https://github.com/terraform-providers/terraform-provider-aws/blob/master/ROADMAP.md#issues-and-enhancements). Given I just provided a review yesterday with only some minor items, it will be under final review/merge when the above items have been addressed or our typical handling of within the next two weeks if there is no response.

**Comment by bflad** [comment]
Hi @gazoakley 👋  Since we didn't hear back from you, we picked up finishing this pull request so it can be released this week. The biggest change not noted in the review above was also including the target type in the resource identifier and lookups since cluster handling is confusing under the hood as RDS will create two targets (one `TRACKING_CLUSTER` and one `RDS_INSTANCE` pointing to the primary cluster instance) and since we do not want to potentially match the wrong one. After that seemed to work well. Thanks so much. 🚀 

Output from acceptance testing in AWS Commercial:

```
--- PASS: TestAccAWSDBProxyTarget_Cluster (567.33s)
--- PASS: TestAccAWSDBProxyTarget_disappears (598.96s)
--- PASS: TestAccAWSDBProxyTarget_Instance (648.21s)
```

Output from acceptance testing in AWS GovCloud (US):

```
--- SKIP: TestAccAWSDBProxyTarget_Instance (2.24s)
--- SKIP: TestAccAWSDBProxyTarget_Cluster (2.24s)
--- SKIP: TestAccAWSDBProxyTarget_disappears (2.24s)
```

**Pull request merged**